Support for astronomical magnitudes

m-wells opened the pull request:
This closes issue #8. I've added unit tests, updated the docs to include usage examples, and updated the readme.  I even handle color index.

mweastwood committed:
initial magnitude tests

mweastwood committed:
experimental support for magnitudes

mweastwood committed:
update require for the latest version of unitful

(this is necessary for log units)

m-wells committed:
added support for AB, bol, Johnson, and Gunn magnitudes

m-wells committed:
merged

m-wells committed:
fixed require

m-wells committed:
updated README

m-wells committed:
color index tests pass

giordano commented on src/UnitfulAstro.jl:
```diff
@@ -86,9 +86,76 @@ import UnitfulAngles: arcminute, arcsecond
+
+# https://en.wikipedia.org/wiki/Color_index
+# for color indices (subtracting magnitudes of different passbands)
+import Base.-
```
All these methods are acts of [type piracy](https://docs.julialang.org/en/v1/manual/style-guide/index.html#Avoid-type-piracy-1), they definitely don't belong to this package, they should be moved to `Unitful.jl`

m-wells replied:
This was inspired by [Unitful.jl itself and how they handle affine quantites](https://github.com/ajkeller34/Unitful.jl/blob/master/src/quantities.jl#L90)
```jl
# Disallow subtracting an affine quantity from a quantity
-(x::AbstractQuantity, y::AffineQuantity) =
    throw(AffineError("an invalid operation was attempted with affine quantities: $x - $y"))
```

giordano replied:
Sure, but these methods don't belong to this package.  They should be defined in `Unitful`, as they refer to `Base` functions with types defined only in that package, not here.

m-wells replied:
These do need to be here (or we move all the  astronomy magnitude business to `Unitful.jl`).  The methods dispatch on `Magnitude`.  The method that allows for color indices to be computed:
```jl
import Base.-
function -(x::Unitful.Level{L,S,T},y::Unitful.Level{L,R,T}) where {L<:Magnitude,S,R,T<:Unitful.Quantity}
    return Unitful.ustrip(x) - Unitful.ustrip(y)
end
```
dispatches only on `Magnitude` log types.  Also notice the `S` and `R`, this permits only different magnitudes to be subtracted (hence the color index).  The other methods here are to allow for magnitudes to be added and subtracted from themselves and to keep people from trying to add differing magnitudes together (you would only subtract different magnitude bands, never add).

giordano replied:
Ok, I missed that `Magnitude` is defined here :grimacing:  So this doesn't make these method type piracy.

> (or we move all the astronomy magnitude business to `Unitful.jl`)

As you have probably seen in the last days, Michael has been a great supporter of adding this feature this to `Unitful.jl` as logarithmic units are heavily used in astronomy.

Still, I wonder if these methods wouldn't be a better fit in `Unitful.jl`.  Do you expect units outside those defined here that wouldn't require these methods?

I don't expect this package to need to define the arithmetic of logarithmic quantities, but only to define the units.

m-wells replied:
So the real kicker here is that astronomers often compute color indices.  For example, if we had two magnitudes `B` and `V` with `B=0.3` and `V=0.9` then we would say `B-V=-0.6`.  This is not what you get if you convert back into `Jy` do the arithmetic and back.  This is a very astronomy thing and I wouldn't expect it to apply anywhere else.

giordano replied:
Ok, fine :slightly_smiling_face:

giordano commented on src/UnitfulAstro.jl:
```diff
@@ -86,9 +86,76 @@ import UnitfulAngles: arcminute, arcsecond
+
+
+struct MagnitudeError <: Exception
+    x
```
Please, do not use [fields with abstract types](https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-fields-with-abstract-type-1).  `String` should be a safe choice for the type of this field

m-wells replied:
This was inspired by [Unitful.jl's AffineError](https://github.com/ajkeller34/Unitful.jl/blob/c01affaa445f0f5307a61cbfe0172d17c45369e9/src/utils.jl#L236)
```jl
"""
    struct AffineError <: Exception
An invalid operation was attempted with affine units / quantities.
"""
struct AffineError <: Exception
    x
end

Base.showerror(io::IO, e::AffineError) = print(io, "AffineError: $(e.x)")
```

giordano replied:
Surely when throwing error performance isn't a big deal.  Still, the way the field is used (only to print the error message) suggests that this field should really be a string, not else.

Most error types in `Base` have the message field marked with a loose `AbstractString`.

m-wells replied:
I can live with `<: AbstractString`

m-wells replied:
I was looking at `Base` and I came across [`showerror`](https://github.com/JuliaLang/julia/blob/8c445663547791e59f33f9e8d49b276332107614/base/errorshow.jl).  I feel like `MagnitudeError` should be typed with `<:Exception`.

giordano replied:
I'm confused, isn't it already an `Exception`? :thinking:

giordano replied:
Oh, [now I see](https://github.com/JuliaAstro/UnitfulAstro.jl/pull/9/commits/aa6674379475ef757a833008e03ea68e7e49a4ac) what you mean.  When i suggested to use `String` or `AbstractString` I was referring to the field (the thread refers to the line of the field and I always talked about the field), not the error itself, which is of course fine with `Exception` :wink:

m-wells replied:
I was going to change it but then thought about it some more . . .

giordano commented on src/UnitfulAstro.jl:
```diff
@@ -86,9 +86,76 @@ import UnitfulAngles: arcminute, arcsecond
+struct MagnitudeError <: Exception
+    x
+end
+Base.showerror(io::IO, e::MagnitudeError) = print(io, "MagnitudeError: $(e.x)")
```
[Avoid string interpolation for I/O](https://docs.julialang.org/en/v1/manual/performance-tips/#Avoid-string-interpolation-for-I/O-1).  There are many case in which interpolation is useful, but here it requires even more characters.
```suggestion
Base.showerror(io::IO, e::MagnitudeError) = print(io, "MagnitudeError: ", e.x)
```

m-wells replied:
This was inspired by [Unitful.jl's AffineError](https://github.com/ajkeller34/Unitful.jl/blob/c01affaa445f0f5307a61cbfe0172d17c45369e9/src/utils.jl#L236)
```jl
"""
    struct AffineError <: Exception
An invalid operation was attempted with affine units / quantities.
"""
struct AffineError <: Exception
    x
end

Base.showerror(io::IO, e::AffineError) = print(io, "AffineError: $(e.x)")
```

giordano replied:
We can do things slightly better :-)

m-wells replied:
How about?
```jl
Base.showerror(io::IO, e::MagnitudeError) = print(io, "MagnitudeError: ", e.x)
```

giordano replied:
Isn't it exactly what I suggested above? :slightly_smiling_face:

m-wells replied:
Ah . . . this is my first time doing a PR this big so I'm not use to doing commits in review like this

giordano commented on src/UnitfulAstro.jl:
```diff
@@ -86,9 +86,76 @@ import UnitfulAngles: arcminute, arcsecond
+@logunit    i_mag   "Gunn i mag"    Magnitude   4760Jy
+@logunit    z_mag   "Gunn z mag"    Magnitude   4810Jy
+
+Unitful.isrootpower_dim(::typeof(Unitful.dimension(AB_mag))) = false
```
I'm pretty sure that also this is type piracy and shouldn't be here.

m-wells replied:
This is how [Unitful.jl handles units](https://github.com/ajkeller34/Unitful.jl/blob/master/src/pkgdefaults.jl#L225)
```jl
# TODO: some more dimensions?
isrootpower_dim(::typeof(dimension(W)))         = false
isrootpower_dim(::typeof(dimension(V)))         = true
isrootpower_dim(::typeof(dimension(A)))         = true
...
```

giordano replied:
Also in this case, `isrrotpower_dim` is a function defined only in that package (unless I'm missing something) and `dimension(AB_mag)` I'm pretty sure is a generic `Unitful` dimension that someone else may be using in their code.

m-wells replied:
`dimension(AB_mag)` is defined in this PR not in `Unitful.jl`

giordano replied:
Precompiling the package gives this warning:
```
WARNING: Method definition isrootpower_dim(Unitful.Dimensions{(Unitful.Dimension{:Mass}(power=Base.Rational{Int64}(num=1, den=1)), Unitful.Dimension{:Time}(power=Base.Rational{Int64}(num=-2, den=1)))}) in module Unitful at /home/mose/.julia/packages/Unitful/W0mMi/src/pkgdefaults.jl:230 overwritten in module UnitfulAstro at /home/mose/.julia/dev/UnitfulAstro/src/UnitfulAstro.jl:114.
```
Doesn't sound quite right.

> `dimension(AB_mag)` is defined in this PR

No, `AB_mag` is defined in this PR, not its dimension:

```julia
julia> dump(Unitful.dimension(UnitfulAstro.AB_mag))
Unitful.Dimensions{(Unitful.Dimension{:Mass}(1//1), Unitful.Dimension{:Time}(-2//1))} 𝐌 𝐓^-2
```

This is by no means a dimension unique to this package.

m-wells replied:
Actually looks like this line isn't need.

giordano replied:
You see? :wink:

giordano reviewed (commented):
Wow, a first PR with full documentation and tests, what a dream!

giordano committed:
Update src/UnitfulAstro.jl

Unnecessary string I/O removed

Co-Authored-By: m-wells <m-wells@users.noreply.github.com>

m-wells committed:
changed MagnitudeError to be of type abstractstring

m-wells committed:
remove undeed isrootpower definition

m-wells committed:
changed MagnitudeError back to Exception

m-wells committed:
removed unneed promotion merging

m-wells commented:
Alright.  I think I addressed all the concerns @giordano had.  I removed two lines that were no longer needed.  Unless there is any other concerns, I think this is ready :smiley:

m-wells commented:
Are we just waiting for @mweastwood to review?

giordano commented:
Yep, I'd like him to have a look, since he's the author and maintainer so far of the package.

mweastwood reviewed (approved):
Apologies for the slow review, my new job doesn't let me spend as much time on Julia as before.

This looks great, thanks for tackling this! The documentation looks really good too.

m-wells commented:
Excellent!  Can we merge now? :smiley:
